From 95b689323b323948a83e1073d8dd105083622809 Mon Sep 17 00:00:00 2001
From: Kai Uwe Broulik <kai.uwe.broulik@basyskom.com>
Date: Mon, 12 Feb 2024 11:10:04 +0100
Subject: [PATCH] QWaylandXdgOutputV1: Send wl_output.done when changing
 properties

When changing logical size or position, only the deprecated
zxdg_output_v1.done event is emitted, if the bound resource uses a version
(i.e. < 3) that still mandates it. However, for version 3+ no done event
is sent unless another change elsewhere causes a wl_output.done event.

The "done" event is supposed to do atomic changes, however, effectively in
Qt Wayland Compositor all changes (except during initialization) are
immediately followed by a "done" event.

Change QWaylandOutput to leave QWaylandXdgOutputV1 alone and switch it
the other way round and let the latter emit wl_output.done if necessary.

Change-Id: Ie37a2241ad950dbd12ecafda055fa6d5712ce05e
---
 .../compositor_api/qwaylandoutput.cpp           | 17 ++++++++---------
 .../compositor_api/qwaylandoutput_p.h           |  4 ++--
 .../extensions/qwaylandxdgoutputv1.cpp          | 15 ++++++---------
 .../extensions/qwaylandxdgoutputv1_p.h          |  1 -
 4 files changed, 16 insertions(+), 21 deletions(-)

diff --git a/src/compositor/compositor_api/qwaylandoutput.cpp b/src/compositor/compositor_api/qwaylandoutput.cpp
index f061afa3..843c5586 100644
--- a/src/compositor/compositor_api/qwaylandoutput.cpp
+++ b/src/compositor/compositor_api/qwaylandoutput.cpp
@@ -38,7 +38,6 @@
 #include <QtWaylandCompositor/private/qwaylandcompositor_p.h>
 #include <QtWaylandCompositor/private/qwaylandview_p.h>
 #include <QtWaylandCompositor/private/qwaylandutils_p.h>
-#include <QtWaylandCompositor/private/qwaylandxdgoutputv1_p.h>
 
 #include <QtCore/QCoreApplication>
 #include <QtCore/QtMath>
@@ -153,9 +152,6 @@ void QWaylandOutputPrivate::sendGeometryInfo()
         if (resource->version() >= 2)
             send_done(resource->handle);
     }
-
-    if (xdgOutput)
-        QWaylandXdgOutputV1Private::get(xdgOutput)->sendDone();
 }
 
 void QWaylandOutputPrivate::sendMode(const Resource *resource, const QWaylandOutputMode &mode)
@@ -179,9 +175,15 @@ void QWaylandOutputPrivate::sendModesInfo()
         if (resource->version() >= 2)
             send_done(resource->handle);
     }
+}
 
-    if (xdgOutput)
-        QWaylandXdgOutputV1Private::get(xdgOutput)->sendDone();
+void QWaylandOutputPrivate::sendDone()
+{
+    const auto values = resourceMap().values();
+    for (auto *resource : values) {
+        if (resource->version() >= 2)
+            send_done(resource->handle);
+    }
 }
 
 void QWaylandOutputPrivate::handleWindowPixelSizeChanged()
@@ -838,9 +840,6 @@ void QWaylandOutput::setScaleFactor(int scale)
     }
 
     Q_EMIT scaleFactorChanged();
-
-    if (d->xdgOutput)
-        QWaylandXdgOutputV1Private::get(d->xdgOutput)->sendDone();
 }
 
 /*!
diff --git a/src/compositor/compositor_api/qwaylandoutput_p.h b/src/compositor/compositor_api/qwaylandoutput_p.h
index d06d1417..e0015730 100644
--- a/src/compositor/compositor_api/qwaylandoutput_p.h
+++ b/src/compositor/compositor_api/qwaylandoutput_p.h
@@ -99,9 +99,9 @@ public:
     void sendMode(const Resource *resource, const QWaylandOutputMode &mode);
     void sendModesInfo();
 
-    void handleWindowPixelSizeChanged();
+    void sendDone();
 
-    QPointer<QWaylandXdgOutputV1> xdgOutput;
+    void handleWindowPixelSizeChanged();
 
 protected:
     void output_bind_resource(Resource *resource) override;
diff --git a/src/compositor/extensions/qwaylandxdgoutputv1.cpp b/src/compositor/extensions/qwaylandxdgoutputv1.cpp
index 2a91e7db..ffb2ff21 100644
--- a/src/compositor/extensions/qwaylandxdgoutputv1.cpp
+++ b/src/compositor/extensions/qwaylandxdgoutputv1.cpp
@@ -156,7 +156,6 @@ void QWaylandXdgOutputManagerV1Private::registerXdgOutput(QWaylandOutput *output
 {
     if (!xdgOutputs.contains(output)) {
         xdgOutputs[output] = xdgOutput;
-        QWaylandOutputPrivate::get(output)->xdgOutput = xdgOutput;
     }
 }
 
@@ -480,7 +479,6 @@ void QWaylandXdgOutputV1Private::sendLogicalPosition(const QPoint &position)
     const auto values = resourceMap().values();
     for (auto *resource : values)
         send_logical_position(resource->handle, position.x(), position.y());
-    needToSendDone = true;
 }
 
 void QWaylandXdgOutputV1Private::sendLogicalSize(const QSize &size)
@@ -488,19 +486,18 @@ void QWaylandXdgOutputV1Private::sendLogicalSize(const QSize &size)
     const auto values = resourceMap().values();
     for (auto *resource : values)
         send_logical_size(resource->handle, size.width(), size.height());
-    needToSendDone = true;
 }
 
 void QWaylandXdgOutputV1Private::sendDone()
 {
-    if (needToSendDone) {
-        const auto values = resourceMap().values();
-        for (auto *resource : values) {
-            if (resource->version() < 3)
-                send_done(resource->handle);
+    const auto values = resourceMap().values();
+    for (auto *resource : values) {
+        if (resource->version() < 3) {
+            send_done(resource->handle);
         }
-        needToSendDone = false;
     }
+
+    QWaylandOutputPrivate::get(output)->sendDone();
 }
 
 void QWaylandXdgOutputV1Private::setManager(QWaylandXdgOutputManagerV1 *_manager)
diff --git a/src/compositor/extensions/qwaylandxdgoutputv1_p.h b/src/compositor/extensions/qwaylandxdgoutputv1_p.h
index ab08f5d4..20f43034 100644
--- a/src/compositor/extensions/qwaylandxdgoutputv1_p.h
+++ b/src/compositor/extensions/qwaylandxdgoutputv1_p.h
@@ -93,7 +93,6 @@ public:
     QSize logicalSize;
     QString name;
     QString description;
-    bool needToSendDone = false;
 
 protected:
     void zxdg_output_v1_bind_resource(Resource *resource) override;
-- 
2.34.1

